### PR TITLE
g_func, dv_func: use nullptr instead of null_gfunc

### LINKED ISSUE #493: Speeding up generate_all_coefficients
[`generate_all_coefficients`](https://github.com/project-asgard/asgard/blob/develop/src/coefficients.cpp#L18) is called several times in IMEX and is the slowest part of the code (per @ckendrick).  This is probably due to the LHS mass matrix is being generated from scratch every time `generate_all_coefficients` is called [link](https://github.com/project-asgard/asgard/blob/develop/src/coefficients.cpp#L36-L45).  This LHS matrix is often the identity matrix, but it's stored as a dense matrix, and then is inverted and left multiplied to the 1D stiffness matrices (thus a dgemm is called).

This should be an easy fix by checking the `g_func_in' and `dv_func_in' inputs against `null_gfunc' and not computing the matrix if so.  

Assigning @quantumsteve.

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

We need some way to differentiate `partial_term<precision>::null_gfunc` from another function. The easiest option is default construct it so that it holds a nullptr.

This fixes #493.

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [ ] Bugfix
- [x] New feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

### CODE REVIEW

File: src/boundary_conditions.cpp [hunk 1]
```diff
@@ -166,6 +165,13 @@ compute_left_boundary_condition(g_func_type<P> const g_func,
   int const level  = dim.get_level();
   int const degree = dim.get_degree();
 
+  if (!g_func)
```
Comment (mkstoyanov) [inline src/boundary_conditions.cpp]: I'm failing to see the idea here? Why replace null gfunc with `partial_term<P>::null_gfunc` which returns `1.0`, I though we wanted to replace `null_gfunc` with checks for `nullptr`.

File: src/boundary_conditions.cpp [hunk 2]
```diff
@@ -166,6 +165,13 @@ compute_left_boundary_condition(g_func_type<P> const g_func,
...
+    g_func = partial_term<P>::null_gfunc;
+  expect(g_func);
```
Comment (mkstoyanov) [inline src/boundary_conditions.cpp]: This can never fail since `g_func` was set in the line above.

File: src/boundary_conditions.cpp [hunk 3]
```diff
@@ -285,7 +297,10 @@ std::vector<fk::vector<P>> generate_partial_bcs(
         dimensions[dim_num].get_degree() *
         fm::two_raised_to(dimensions[dim_num].get_level());
     auto const &f = bc_funcs[dim_num];
-    auto const &g = terms[dim_num].get_partial_terms()[p_index].g_func;
+    auto g        = terms[dim_num].get_partial_terms()[p_index].g_func;
+    if (!g)
+      g = partial_term<P>::null_gfunc;
```
Comment (mkstoyanov) [inline src/boundary_conditions.cpp]: Another expect `g` and setting `g` to `null_gfunc`

**Review by mkstoyanov — CHANGES_REQUESTED** [review]
Maybe I'm missing the larger picture, is this meant to be part of a bigger change?

File: src/coefficients.cpp [hunk 4]
```diff
@@ -50,13 +49,14 @@ void generate_all_coefficients(
...
         for (int level = 0; level <= pde.max_level; ++level)
         {
-          auto const dof = dim.get_degree() * fm::two_raised_to(level);
-          auto mass_tmp  = mass_coeff.extract_submatrix(0, 0, dof, dof);
-
           auto result = generate_coefficients<P>(
               dim, partial_terms[k], transformer, level, time, rotate);
-
-          fm::gesv(mass_tmp, result, ipiv);
+          if (partial_terms[k].dv_func || partial_terms[k].g_func)
```
Comment (quantumsteve) [inline src/coefficients.cpp:54]: @mkstoyanov Here's where I'm checking if `partial_terms[k].dv_func` or `partial_terms[k].g_func` are defined and skipping the call to `fm::gesv` if both return `false`.

File: src/boundary_conditions.cpp (hunk 1 above)
Comment (quantumsteve) [reply src/boundary_conditions.cpp]: I've now fully removed `partial_term<P>::null_gfunc` from ASGarD.

File: src/boundary_conditions.cpp (hunk 2 above)
Comment (quantumsteve) [reply src/boundary_conditions.cpp]: removed

File: src/boundary_conditions.cpp (hunk 3 above)
Comment (quantumsteve) [reply src/boundary_conditions.cpp]: I've now fully removed `partial_term<P>::null_gfunc` from ASGarD.

**Review by mkstoyanov — DISMISSED** [review]

**Review by mkstoyanov — APPROVED** [review]

**Pull request merged**